src/java.base/share/classes/java/lang/StringUTF16.java

```diff
@@ -208,30 +392,23 @@ public static int compress(byte[] src, int srcOff, byte[] dst, int dstOff, int l
         return len;
     }
 
+    // Create the UTF16 buffer for !COMPACT_STRINGS
     public static byte[] toBytes(int[] val, int index, int len) {
         final int end = index + len;
-        // Pass 1: Compute precise size of char[]
-        int n = len;
-        for (int i = index; i < end; i++) {
-            int cp = val[i];
-            if (Character.isBmpCodePoint(cp))
-                continue;
-            else if (Character.isValidCodePoint(cp))
-                n++;
-            else throw new IllegalArgumentException(Integer.toString(cp));
-        }
-        // Pass 2: Allocate and fill in <high, low> pair
+        int n = computeCodePointSize(val, index, end);
+
         byte[] buf = newBytesFor(n);
-        for (int i = index, j = 0; i < end; i++, j++) {
-            int cp = val[i];
-            if (Character.isBmpCodePoint(cp)) {
-                putChar(buf, j, cp);
-            } else {
-                putChar(buf, j++, Character.highSurrogate(cp));
-                putChar(buf, j, Character.lowSurrogate(cp));
-            }
-        }
-        return buf;
+        return extractCodepoints(val, index, len, buf, 0);
+     }
+
+    // Return the number of chars needed to store the code point.
+    private static int charsPerCodePoint(int codePoint) {
+        if (Character.isBmpCodePoint(codePoint)) {
+            return 1;
+        } else if (Character.isValidCodePoint(codePoint)) {
+            return 2;
+        } else
+            throw new IllegalArgumentException(Integer.toString(codePoint));
```

Comment:
`toHexString` might be more informative.

Reply:
Perhaps, but changing the exception text is out scope for this PR; it has been decimal since JDK 9 (2015).